packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -0,0 +1,250 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const error = require('../error');
+const subhub = require('../subhub/client');
+
+const stripe = require('stripe');
+
+/** @typedef {import('stripe').customers.ICustomer} Customer */
+/** @typedef {import('stripe').products.IProduct} Product */
+/** @typedef {import('stripe').plans.IPlan} Plan */
+/** @typedef {import('stripe').subscriptions.ISubscription} Subscription */
+/** @typedef {import('stripe').customers.ICustomerSubscriptions} Subscriptions */
+
+/**
+ * @typedef AbbrevProduct
+ * @property {string} product_id
+ * @property {string} product_name
+ * @property {Product['metadata']} product_metadata
+ */
+
+/**
+ * @typedef AbbrevPlan
+ * @property {string} plan_id
+ * @property {string} plan_name
+ * @property {Product['metadata']} plan_metadata
+ * @property {string} product_id
+ * @property {string} product_name
+ * @property {Product['metadata']} product_metadata
+ * @property {Plan['interval']} interval
+ * @property {Plan['amount']} amount
+ * @property {Plan['currency']} currency
+ */
+
+/**
+ * Get a cached result at a cache key and regenerated it with `refreshFunction`
+ * if its expired.
+ *
+ * @template T
+ * @param {*} redis
+ * @param {string} cacheKey
+ * @param {number} cacheTtl
+ * @param {() => Promise<T>} refreshFunction
+ * @returns {Promise<T>} possibly cached result
+ */
+async function cachedResult(log, redis, cacheKey, cacheTtl, refreshFunction) {
+  if (cacheTtl) {
+    try {
+      const json = await redis.get(cacheKey);
+      if (json) {
+        return JSON.parse(json);
+      }
+    } catch (err) {
+      log.error(
+        `subhub.cachedResult.${refreshFunction.name}.getCachedResponse.failed`,
```

Comment:
What is the name if an anonymous function is passed in?

Reply:
It's `undefined`. I'll add a note to this function that one should only pass in functions with a name for useful errors.